mm.c

```diff
@@ -118,9 +124,36 @@ static size_t find_free_list_index(size_t words) {
 	physical memory with neigboring free blocks.
 	Returns the pointer to the beginning of this
 	new free block.
+	Coalesce is only called when a taken block is freed,
+	before the free block is placed into the free list.
+	As such, coalesce does not set pointer values.
 */
 static void *coalesce(void *bp) {
-
+	char **prev_block = PREV_BLOCK_IN_MEM(bp);
+	char **next_block = NEXT_BLOCK_IN_MEM(bp);
+	size_t prev_status = GET_STATUS(prev_block);
+	size_t next_status = GET_STATUS(next_block);
+	size_t new_size = GET_SIZE(bp);
+
+	if (prev_status == TAKEN && next_status == TAKEN) {
+		return bp;
+	} else if (prev_status == TAKEN && next_status == FREE) {
+		new_size += GET_SIZE(next_block) + HDR_FTR_SIZE;
```

Comment:
The current block, no, because coalesce is only called once a TAKEN block becomes FREE.

The prev/next blocks, you are right about. We need to make sure that the block is taken out of the free list. 

Do you think it will be a good idea to use `alloc_free_block` to achieve this task? The function does exactly that (removes the block from the free list) and then `coalesce` changes the TAKEN status to FREE.

This will reduce code redundancy, and also that way until `alloc_free_block` is implemented, `coalesce` will still work.